File: FeatureDemos/C++/AnonymousMethods/ModuleMain.cpp

//---------------------------------------------------------------------------
#include "ModuleMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"

TComponentClass WebModuleClass = __classid(TWebModuleMain);
//---------------------------------------------------------------------------

//---------------------------------------------------------------------------
// TWebStencilsLookupFuncWrapper implementation
//---------------------------------------------------------------------------
bool __fastcall TWebStencilsLookupFuncWrapper::Invoke(TWebStencilsDataVar* AVar, const System::UnicodeString APropName, System::UnicodeString &AValue)
{
  return FFunc ? FFunc(AVar, APropName, AValue) : false;
}

//---------------------------------------------------------------------------
// TWebModuleMain implementation
//---------------------------------------------------------------------------
__fastcall TWebModuleMain::TWebModuleMain(TComponent* Owner)
  : TWebModule(Owner)
{
}
//---------------------------------------------------------------------------

__fastcall TWebModuleMain::~TWebModuleMain()
{
}
//---------------------------------------------------------------------------

void __fastcall TWebModuleMain::WebModuleMainDefaultHandlerAction(TObject *Sender, TWebRequest *Request,
          TWebResponse *Response, bool &Handled)
{
  // Example 1: Using a lambda with an object (TDictionary)
  // This is equivalent to Delphi's env1 with anonymous method + TDictionary
  // Create dictionary for this request (WebStencils will own it when AOwned=true)
  auto* dict = new System::Generics::Collections::TDictionary__2<System::UnicodeString, System::UnicodeString>();
  dict->Add("APP_VERSION", "1.0.0");
  dict->Add("DEBUG_MODE", "True");
  
  // Create a lambda that accesses the dictionary
  // Capture dict by value in the lambda since WebStencils will own it
  auto lookupFunc1 = MakeLookupFunc(
    [dict](TWebStencilsDataVar* AVar, const System::UnicodeString& APropName, System::UnicodeString& AValue) -> bool
    {
      System::UnicodeString UpperPropName = APropName.UpperCase();
      return dict->TryGetValue(UpperPropName, AValue);
    }
  );
  
  WSProcessor->AddVar("env1", dict, true, lookupFunc1);
  
  // Example 2: Fully anonymous lambda without an associated object
  // This is equivalent to Delphi's env2 with fully anonymous method
  auto lookupFunc2 = MakeLookupFunc(
    [](TWebStencilsDataVar* AVar, const System::UnicodeString& APropName, System::UnicodeString& AValue) -> bool
    {
      System::UnicodeString UpperPropName = APropName.UpperCase();
      if (UpperPropName == "APP_VERSION")
      {
        AValue = "1.0.0";
        return true;
      }
      else if (UpperPropName == "DEBUG_MODE")
      {
        AValue = "True";
        return true;
      }
      return false;
    }
  );
  
  WSProcessor->AddVar("env2", nullptr, false, lookupFunc2);
  
  Response->Content = WSProcessor->Content();
}
//---------------------------------------------------------------------------

File: FeatureDemos/Delphi/AnonymousMethods/Module.Main.pas
unit Module.Main;

interface

uses
  System.SysUtils, System.Classes, Web.HTTPApp, System.Generics.Collections, Web.Stencils;

type

  TMap = TDictionary<string, string>;

  TWebModuleMain = class(TWebModule)
    WSProcessor: TWebStencilsProcessor;
    procedure WebModuleMainDefaultHandlerAction(Sender: TObject;
      Request: TWebRequest; Response: TWebResponse; var Handled: Boolean);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  WebModuleClass: TComponentClass = TWebModuleMain;

implementation

{%CLASSGROUP 'System.Classes.TPersistent'}

{$R *.dfm}

procedure TWebModuleMain.WebModuleMainDefaultHandlerAction(Sender: TObject;
  Request: TWebRequest; Response: TWebResponse; var Handled: Boolean);
var
  LDict: TMap;
begin
  LDict := TMap.Create;
  LDict.Add('APP_VERSION', '1.0.0');
  LDict.Add('DEBUG_MODE', 'True');

  // Example 1: Using anonymous method with an object (TDictionary)
  // The anonymous method accesses properties from the associated object
  WSProcessor.AddVar('env1', LDict, True,
    function(AVar: TWebStencilsDataVar; const APropName: string; var AValue: string): Boolean
    begin
      Result := TMap(AVar.TheObject).TryGetValue(APropName.ToUpper, AValue);
    end);

  // Example 2: Fully anonymous method without an associated object
  // The anonymous method provides values directly without an object reference
  WSProcessor.AddVar('env2', nil, false,
    function (AVar: TWebStencilsDataVar; const APropName: string; var AValue: string): Boolean
    begin
      if APropName.ToUpper = 'APP_VERSION' then
        AValue := '1.0.0'
      else if APropName.ToUpper = 'DEBUG_MODE' then
        AValue := 'True'
      else
      begin
        Result := False;
        Exit;
      end;
      Result := True;
    end);

  Response.Content := WSProcessor.Content;
end;

end.

